File: java-factorial/src/main/java/isa/factorial/design_patterns/FactorialImplSingleton.java
package isa.factorial.design_patterns;

import isa.factorial.Factorial;

public class FactorialImplSingleton implements Factorial {

    private static FactorialImplSingleton instance;
    
    private FactorialImplSingleton() {

    }

    public static FactorialImplSingleton getInstance() {
        if (instance == null) {
            instance = new FactorialImplSingleton();
        }
        return instance;
    }

    @Override
    public long fact(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("Error: n negativo. El número debe ser >= 0.");
        }

        return (n > 0) ? n * fact(n - 1) : 1;
    }
}

File: python-factorial/factorial/design_patterns/factorial_impl_singleton.py
from factorial import Factorial

class MetaSingleton(type):
    _instances = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(MetaSingleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]
    
    
class FactorialImpl(Factorial, metaclass=MetaSingleton):
   
    def fact(self, n: int) -> int:
        if not isinstance(n, int):
            raise TypeError('Error: n no es un número entero.')
        if n < 0:
            raise ValueError('Error: n negativo. El número n debe ser positivo.')
        return n * self.fact(n-1) if n > 0 else 1
